contracts/sol6/KyberFprReserveVBSC.sol

```diff
@@ -0,0 +1,431 @@
+/// for working with weth should specify external address to hold weth.
+/// Allow Reserve to set maxGasPriceWei to trade with
+contract KyberFprReserveVBSC is
+    IKyberReserve,
+    Utils,
+    Withdrawable,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20Ext;
+    using SafeMath for uint256;
+
+    mapping(bytes32 => bool) public approvedWithdrawAddresses; // sha3(token,address)=>bool
+    mapping(address => address) public tokenWallet;
+
+    struct ConfigData {
+        bool tradeEnabled;
+        uint128 maxGasPriceWei;
+    }
+
+    ConfigData internal configData;
+
+    IConversionRates public conversionRatesContract;
+    IKyberSanity public sanityRatesContract;
+    IWeth public weth;
+    IERC20Ext public quoteToken;
+    
+
+    event DepositToken(IERC20Ext indexed token, uint256 amount);
+    event TradeExecute(
+        address origin,
+        address indexed trader,
+        IERC20Ext indexed src,
+        uint256 srcAmount,
+        IERC20Ext indexed destToken,
+        uint256 destAmount,
+        address payable destAddress
+    );
+    event TradeEnabled(bool enable);
+    event MaxGasPriceUpdated(uint128 newMaxGasPrice);
+    event WithdrawAddressApproved(
+        IERC20Ext indexed token,
+        address indexed addr,
+        bool approve
+    );
+    event NewTokenWallet(IERC20Ext indexed token, address indexed wallet);
+    event WithdrawFunds(
+        IERC20Ext indexed token,
+        uint256 amount,
+        address indexed destination
+    );
+    event SetConversionRateAddress(IConversionRates indexed rate);
+    event SetWethAddress(IWeth indexed weth);
+    event SetSanityRateAddress(IKyberSanity indexed sanity);
+
+    constructor(
+        IConversionRates _ratesContract,
+        IWeth _weth,
+        IERC20Ext _quoteToken,
+        uint128 _maxGasPriceWei,
+        address _admin
+    ) public Withdrawable(_admin) {
+        require(_ratesContract != IConversionRates(0), "ratesContract 0");
+        require(_weth != IWeth(0), "weth 0");
+        conversionRatesContract = _ratesContract;
+        weth = _weth;
+        quoteToken = _quoteToken;
+        configData = ConfigData({
+            tradeEnabled: true,
+            maxGasPriceWei: _maxGasPriceWei
+        });
+    }
+
+    receive() external payable {
+        emit DepositToken(ETH_TOKEN_ADDRESS, msg.value);
+    }
+
+    function trade(
+        IERC20Ext srcToken,
+        uint256 srcAmount,
+        IERC20Ext destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool /* validate */
+    ) public payable override nonReentrant  returns (bool) {
+        ConfigData memory data = configData;
+        require(data.tradeEnabled, "trade not enable");
+        require(
+            tx.gasprice <= uint256(data.maxGasPriceWei),
+            "gas price too high"
+        );
+
+        doTrade(srcToken, srcAmount, destToken, destAddress, conversionRate);
+
+        return true;
```

Comment:
Can this be changed to how network `tradeWithHintAndFee` works, where return value is `uint256 destAmount`? Relevant to Krystal integration in BSC.